xmtp_mls/src/groups/message_history.rs

```diff
@@ -178,37 +270,158 @@ where
         Ok(())
     }
 
-    pub(crate) fn provide_pin(&self, pin_challenge: &str) -> Result<(), GroupError> {
-        let conn = self.store().conn()?;
-        let sync_group_id = conn
-            .find_sync_groups()?
-            .pop()
-            .ok_or(GroupError::GroupNotFound)?
-            .id;
+    pub async fn get_pending_history_request(
+        &self,
+    ) -> Result<Option<(String, String)>, MessageHistoryError> {
+        let sync_group = self.get_sync_group()?;
+
+        // sync the group
+        sync_group.sync(self).await?;
 
-        let requests = conn.get_group_messages(
-            sync_group_id,
+        let messages = sync_group.find_messages(
+            Some(GroupMessageKind::Application),
+            None,
+            None,
             None,
             None,
+        )?;
+        let last_message = messages.last();
+        let history_request: Option<(String, String)> = match last_message {
+            Some(msg) => {
+                let message_history_content =
+                    serde_json::from_slice::<MessageHistoryContent>(&msg.decrypted_message_bytes);
+                match message_history_content {
+                    // if the last message is a request, return its request ID and pin code
+                    Ok(MessageHistoryContent::Request(request)) => {
+                        Some((request.request_id, request.pin_code))
+                    }
+                    _ => None,
+                }
+            }
+            None => None,
+        };
+
+        Ok(history_request)
+    }
+
+    pub async fn reply_to_history_request(
+        &self,
+    ) -> Result<MessageHistoryReply, MessageHistoryError> {
+        let pending_request = self.get_pending_history_request().await?;
+
+        if let Some((request_id, _)) = pending_request {
```

Comment:
How do i know which request I'm replying to? What happens if a recent request just snuck in

Reply:
this will reply the latest request message in the sync group, which is sync'd prior to checking. does that make sense?

Reply:
I think that's fine for now. Just want to walk through a scenario.

1. AppB receives a message history request from AppA and shows it to the user. They look at it and click OK.
2. App receives a second message history request from App C
3. libxmtp replies to the request from App C because it's the newest request at the time of reply

We can solve that issue later, but it's something we should write a ticket for